src/Spectre.libGaussianMixtureModelling.Tests/ExpectationMaximizationTest.cpp

```diff
@@ -0,0 +1,287 @@
+
+        GaussianMixtureModel result = em.EstimateGmm();
+        std::vector<double> approximates = GenerateValues(mzs, result.components);
+
+        std::vector<double> errors(intensities.size());
+        for (unsigned i = 0; i < intensities.size(); i++)
+        {
+            errors[i] = pow(intensities[i] - approximates[i], 2) * 10000.0; // scale the errors
+        }
+
+        double averageError = 0.0;
+        for (unsigned i = 0; i < errors.size(); i++)
+        {
+            averageError += errors[i];
+        }
+        averageError /= errors.size();
+        double maxError = *std::max_element(errors.begin(), errors.end());
+        double minError = *std::min_element(errors.begin(), errors.end());
+        printf("Avg error: %f\n", averageError);
+        printf("Max error: %f\n", maxError);
+        printf("Min error: %f\n", minError);
+
+        EXPECT_LE(averageError, 0.1);
+    }
+
+    TEST_F(ExpectationMaximizationTest, test_em_ref_initialization)
+    {
+        std::mt19937_64 rngEngine(0);
+        std::vector<GaussianComponent> components(gaussianComponents.size());
+        BasicInitializationRef initialization(&mzs[0], (unsigned)mzs.size(),
+            components, rngEngine);
+
+        // Check if values of assigned means come from available mz values
+        initialization.AssignRandomMeans();
+
+        for (int i = 0; i < components.size(); i++)
+        {
+            EXPECT_NE(std::find(mzs.begin(), mzs.end(), components[i].mean), mzs.end());
+        }
+
+        // Calculate sample variance
+        initialization.AssignVariances();
+
+        double sampleMean = std::accumulate(mzs.begin(), mzs.end(), 0.0) / (double)mzs.size();
+        double variance = 0.0;
+        for (unsigned i = 0; i < mzs.size(); i++)
+        {
+            variance += pow(mzs[i] - sampleMean, 2);
+        }
+        variance /= (double)mzs.size();
+        double deviation = sqrt(variance);
+
+        // Check if sample variance was properly assigned to all components
+        for (unsigned i = 0; i < components.size(); i++)
+        {
+            EXPECT_EQ(components[i].deviation, deviation);
+        }
+
+        // Check if weights have been assigned uniformly
+        initialization.AssignWeights();
+
+        double weight = 1.0 / (double)components.size();
+        for (unsigned i = 0; i < components.size(); i++)
+        {
+            EXPECT_EQ(components[i].weight, weight);
+        }
+    }
+
+    TEST_F(ExpectationMaximizationTest, test_em_ref_expectation)
+    {
+        DataType** affilationMatrix = new DataType*[mzs.size()];
+        affilationMatrix[0] = new DataType[gaussianComponents.size() * mzs.size()];
+        for (unsigned i = 1; i < mzs.size(); ++i)
+            affilationMatrix[i] = affilationMatrix[0] + i * gaussianComponents.size();
+
+        ExpectationRunnerRef expectation(&mzs[0], (unsigned)mzs.size(), 
+            (DataType**&)affilationMatrix, gaussianComponents);
```

Comment:
I knew this will be questioned :D. The following code explains this phenomena:
https://github.com/spectre-team/spectre/pull/95/files#diff-e68f3d339f4421e13b7d9f3a0b6f15c5R56

Since the expectation step will update the affiliation matrix, it needs a reference to this pointer, Passing the pointer itself wouldn't work, since it is initialized in the constructor body, after the initialization list have finished its execution.